### PR TITLE
Adding PushNotification{Enqueued/WillBeSent/HasBeenSent} plugin hooks

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Adding PushNotification{Enqueued/WillBeSent/HasBeenSent} plugin hooks

This will allow us to use the plugin
github.com/jespino/mattermost-plugin-push-notifications-recorder to store data
about push notifications and investigate the delays in notifications

#### Ticket Link
[MM-14299](https://mattermost.atlassian.net/browse/MM-14299)

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)

### CODE REVIEW

**Comment by levb** [comment]
(added 2 more reviewers - to agree on API changes)

File: plugin/hooks.go [hunk 1]
```diff
@@ -149,4 +153,16 @@ type Hooks interface {
...
+	PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification
+
+	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
+	PushNotificationHasBeenSent(c *Context, notification *model.PushNotification)
+
+	// PushNotificationHasFailed is invoked after the push proxy server replies with an error.
+	PushNotificationHasFailed(c *Context, notification *model.PushNotification, err error)
+
+	// PushNotificationEnqueued is invoked just before the notification has been enqueued to be sent to the push proxy server.
+	PushNotificationEnqueued(c *Context, notificationId, notificationType, userId, channelId, postId string)
```
Comment (hanzei) [inline plugin/hooks.go]: Could you elaborate a bit why this is needed? Why isn't the parameter of this function `*model.PushNotification`?

File: plugin/hooks.go [hunk 2]
```diff
@@ -149,4 +153,16 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
+	PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification
+
+	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
```
Comment (hanzei) [inline plugin/hooks.go]: :point_up_2: 

```suggestion
	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
	//  Minimum server version: 5.10
```

File: plugin/hooks.go [hunk 3]
```diff
@@ -149,4 +153,16 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
```
Comment (hanzei) [inline plugin/hooks.go]: Could you please document the minimum server version:
```suggestion
	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
	// Minimum server version: 5.10
```

Comment (hanzei) [inline plugin/hooks.go]: The documentation is a bit short, especially about the return value. Please document two things:
- Plugins can modify the notification by changing the return value
- Plugins can reject the notification by returning `nil`

File: plugin/hooks.go [hunk 4]
```diff
@@ -149,4 +153,16 @@ type Hooks interface {
...
+	PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification
+
+	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
+	PushNotificationHasBeenSent(c *Context, notification *model.PushNotification)
+
+	// PushNotificationHasFailed is invoked after the push proxy server replies with an error.
```
Comment (hanzei) [inline plugin/hooks.go]: Wording change for consistency

```suggestion
	// PushNotificationHasFailed is invoked after the push proxy server has replied with an error.
```

File: plugin/hooks.go (hunk 1 above)
Comment (jespino) [reply plugin/hooks.go]: The reason behind this, is because the push notifications are enqueued in a pull of workers that are who really send the notifications to the push proxy. The notifications itself are generated by the worker, so, Before I enqueue the notification I don't have the entire notification.

Comment (hanzei) [reply plugin/hooks.go]: This seems to be a very "internal" hook. I can't think of a plugin (other then your notification plugin), that would use this hook. I suppose must plugin devs would just use `PushNotificationWillBeSent`.

I'm hesitant so expose a lot of hooks, because we have to support and maintain them. 1/5 on not adding this as a "public hook".

What do you think about labels this as "internal" and warn about us changing it any time? Aren't there some API endpoints where we do the same?

Comment (jespino) [reply plugin/hooks.go]: I'm ok about labeling it as "internal", and do not include that in the plugins documentation (or include it but clearly explaining that is not inside the backward compatibility policy).

File: plugin/hooks.go (hunk 3 above)
Comment (jespino) [reply plugin/hooks.go]: Done

Comment (jespino) [reply plugin/hooks.go]: Done

File: plugin/hooks.go (hunk 2 above)
Comment (jespino) [reply plugin/hooks.go]: Done

File: plugin/hooks.go (hunk 3 above)
Comment (jespino) [reply plugin/hooks.go]: Done

File: plugin/hooks.go (hunk 4 above)
Comment (jespino) [reply plugin/hooks.go]: Done

File: plugin/hooks.go [hunk 5]
```diff
@@ -149,4 +154,41 @@ type Hooks interface {
...
+	PushNotificationHasBeenSent(c *Context, notification *model.PushNotification)
+
+	// PushNotificationHasFailed is invoked after the push proxy server has replied with an error.
+	//
+	// Minimum server version: 5.10
+	PushNotificationHasFailed(c *Context, notification *model.PushNotification, err error)
+
+	// PushNotificationEnqueued is invoked just before the notification has been enqueued to be sent to the push proxy server.
+	//
+	// This is an internal API and it can change in the future.
```
Comment (hanzei) [inline plugin/hooks.go]:
```suggestion
	// This is an internal hook and it can change anytime in the future.
```

File: plugin/hooks.go [hunk 6]
```diff
@@ -149,4 +154,41 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
+	// You should return the notification that you want to be sent to the user (normally the same notification).
+	//
+	// You can reject the notification returning a nil notification.
+	//
+	// You can modify the notification, returning a changed version of the notification.
```
Comment (hanzei) [inline plugin/hooks.go]: For consistency:

```suggestion
	// You can modify the notification, returning a changed version of the notification.
	//
	//  If you don't need to modify or rejected notifications, use PushNotificationWillBeSent instead. 
```

File: plugin/hooks.go [hunk 7]
```diff
@@ -149,4 +154,41 @@ type Hooks interface {
...
+	//
+	// This is an internal API and it can change in the future.
+	//
+	// Minimum server version: 5.10
+	PushNotificationEnqueued(c *Context, notificationId, notificationType, userId, channelId, postId string)
+
+	// PushNotificationAck is invoked when the server receive the confirmation
+	// of the reception of the push notification from the mobile device.
+	//
+	// You can return data that will be return to the mobile device as response to the ACK.
```
Comment (hanzei) [inline plugin/hooks.go]: Could you give an example on what kind of data could be returned?

Is this really something plugin authors can modify, as the client would need to know what do with the data?

File: plugin/hooks.go [hunk 8]
```diff
@@ -149,4 +154,41 @@ type Hooks interface {
...
+	//
+	// Or you can return an error that will be returned to the mobile device as response to the ACK.
```
Comment (hanzei) [inline plugin/hooks.go]: What happens when the mobile device receives an error?

Comment (jespino) [reply plugin/hooks.go]: Probably it will retry after X time, or so. I'm not sure, it depends on the mobile implementation. And probably if you return data, can be a modified version of the notification to update it in the device.

Comment (jespino) [reply plugin/hooks.go]: I need to talk with @enahum to be sure what we expect here. Probably we will merge it when the whole thing is ready (the mobile apps code too).

File: plugin/hooks.go (hunk 6 above)
Comment (jespino) [reply plugin/hooks.go]: You meant `PushNotificationHasBeenSent`? It is not exactly the same because `WillBeSent` is executed every time a push notification is sent to the push proxy, but `HasBeenSent` is only executed if the notification has been sent without errors. It depends on your intentions with the plugin.

File: plugin/hooks.go (hunk 7 above)
Comment (jespino) [reply plugin/hooks.go]: I going to change it to a notification, because is the unique data type that makes sense. If the client receive a notification from here, update the current notification in the app (or at least, the interanal representation, to for example, open the correct channel and post if you have removed it from the original notification). This is because this way you can make a plugin which remove every single bit of internal information from the notification (so this way is not passing through google/apple servers, and only when it arrives the other end, you can get the information from the server using this ACK endpoint).

File: app/notification_push.go [hunk 9]
```diff
@@ -294,11 +359,40 @@ func (a *App) sendToPushProxy(msg model.PushNotification, session *model.Session
 		mlog.Info(fmt.Sprintf("Device was reported as removed for UserId=%v SessionId=%v removing push for this session", session.UserId, session.Id), mlog.String("user_id", session.UserId))
 		a.AttachDeviceId(session.Id, "", session.ExpiresAt)
 		a.ClearSessionCacheForUser(session.UserId)
+		err := errors.New("Device was reported as removed")
```
Comment (levb) [inline app/notification_push.go]: nit: maybe a little in-line function for the 3 similar flows here?

File: app/notification_push.go [hunk 10]
```diff
@@ -270,20 +306,49 @@ func (a *App) StopPushNotificationsHubWorkers() {
...
+	if updatedMsg == nil {
+		return nil
+	}
+
+	request, err := http.NewRequest("POST", strings.TrimRight(*a.Config().EmailSettings.PushNotificationServer, "/")+model.API_URL_SUFFIX_V1+"/send_push", strings.NewReader(updatedMsg.ToJson()))
 	if err != nil {
-		mlog.Error(fmt.Sprintf("Error sending to push proxy: UserId=%v SessionId=%v message=%v",
-			session.UserId, session.Id, err.Error()), mlog.String("user_id", session.UserId))
-		return
+		if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
```
Comment (levb) [inline app/notification_push.go]: 1/5 for this failure (failed to compose an HTTP request) invoking the plugin is an overkill.

File: api4/system.go [hunk 11]
```diff
@@ -515,3 +517,38 @@ func getRedirectLocation(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(model.MapToJson(m)))
 	return
 }
+
+func pushNotificationsAck(c *Context, w http.ResponseWriter, r *http.Request) {
```
Comment (levb) [inline api4/system.go]: Sorry I don't understand the flow, but where is this ack coming from?

Comment (jespino) [reply api4/system.go]: This comes from the mobile app, the idea is, this will be implemented in the mobile app and, if the server configuration enables PushNotificationAck, it will send an Ack message as soon as it can. This way we can confirm the entire flow of the push notification.

File: app/notification_push.go (hunk 10 above)
Comment (jespino) [reply app/notification_push.go]: Is not about the failure composing the HTTP request, is about the failure trying to send the notification. If I try to send the notification and for any reason is not sent, I need to know from the plugins side (If we want to have the possibility of monitor entirely the push notification process).

File: app/notification_push.go (hunk 9 above)
Comment (jespino) [reply app/notification_push.go]: Yes, good point, I going to rewrite it

Comment (jespino) [reply app/notification_push.go]: Done

File: api4/system.go (hunk 11 above)
Comment (levb) [reply api4/system.go]: 0/5 and may be asking too much even if it makes sense, but... should the DoAck/handleAck be implemented by the notification tracking plugins? I understand that installing the necessary hook in the client may not be easily possible, but wouldn't that be "The Right Thing to Do"?

File: app/notification_push.go (hunk 10 above)
Comment (levb) [reply app/notification_push.go]: These should practically never happen anyway. So as long as the boilerplate code is compacted I don't mind.

File: api4/system.go (hunk 11 above)
Comment (jespino) [reply api4/system.go]: The thing is we can wait until we can create client plugins in the mobile apps, but anyway makes sense to provide the entire push notification flow in this case. Actually this way we can implement other kind of plugins, not only for monitoring. We can implement anonymization of notifications (like, remove any private/sensitive data from the message, and give it to the user using the ACK, avoiding the data to travel through apple and google servers). We can implement "push notification enrichment", like add during the ack attachments to the notification, and show it (the push notification itself have some limitations of size).

Basically, looks like a good thing to have out of the box, without the need of extra plugin for client, and we really need a better way to understand how notifications are behaving right now, and I think we can't wait until mobile plugins are available.

File: api4/system.go [hunk 12]
```diff
@@ -515,3 +517,38 @@ func getRedirectLocation(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	var err *model.AppError
+	var notification *model.PushNotification
+	ackAt := model.GetMillis()
+
+	if pluginsEnvironment := c.App.GetPluginsEnvironment(); pluginsEnvironment != nil {
+		pluginContext := c.App.PluginContext()
+		pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			notification, err = hooks.PushNotificationAck(pluginContext, ack.Id, ack.DeviceId, ack.ReceivedAt, ackAt)
```
Comment (enahum) [inline api4/system.go]: Why do you need the device ID? the ack.Id should be enough

File: api4/system.go [hunk 13]
```diff
@@ -515,3 +517,38 @@ func getRedirectLocation(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		mlog.Debug(fmt.Sprintf("Push notification %s ack: message recived in the device %s at %d, and ack received at server at %d, but failed with the error %v", ack.Id, ack.DeviceId, ack.ReceivedAt, ackAt, err))
+		c.Err = err
+		return
+	}
+
+	mlog.Debug(fmt.Sprintf("Push notification %s ack: message recived in the device %s at %d, and ack received at server at %d", ack.Id, ack.DeviceId, ack.ReceivedAt, ackAt))
+
+	if notification != nil {
+		w.Write([]byte(notification.ToJson()))
```
Comment (enahum) [inline api4/system.go]: why do we need to return this in the response?

File: app/notification_push.go [hunk 14]
```diff
@@ -132,8 +140,20 @@ func (a *App) sendPushNotification(notification *postNotification, user *model.U
 	channelName := notification.GetChannelName(nameFormat, user.Id)
 	senderName := notification.GetSenderName(nameFormat, *cfg.ServiceSettings.EnablePostUsernameOverride)
 
+	notificationId := model.NewId()
```
Comment (enahum) [inline app/notification_push.go]: you are already receiving this id as a parameter, correct?

File: app/notification_push.go [hunk 15]
```diff
@@ -212,13 +233,27 @@ func (a *App) ClearPushNotificationSync(userId string, channelId string) {
...
-		a.sendToPushProxy(tmpMessage, session)
+		err := a.sendToPushProxy(tmpMessage, session)
+		if err != nil {
+			mlog.Error(fmt.Sprintf("Error sending push notification: UserId=%v SessionId=%v message=%v", session.UserId, session.Id, err.Error()), mlog.String("user_id", session.UserId))
+		}
 	}
 }
 
 func (a *App) ClearPushNotification(userId string, channelId string) {
+	notificationId := model.NewId()
```
Comment (enahum) [inline app/notification_push.go]: this is tricky, including an id in the clear notification should be use to clear a specific notification. see https://github.com/wix/react-native-notifications/blob/master/docs/advancedIos.md

But I guess we want to use it for something else, so lets leave it as is

File: plugin/hooks.go [hunk 16]
```diff
@@ -149,4 +154,43 @@ type Hooks interface {
...
+	//
+	// This is an internal hook and it can change anytime in the future.
+	//
+	// Minimum server version: 5.10
+	PushNotificationEnqueued(c *Context, notificationId, notificationType, userId, channelId, postId string)
+
+	// PushNotificationAck is invoked when the server receive the confirmation
+	// of the reception of the push notification from the mobile device.
+	//
+	// You can return a notification that will be return to the mobile device
```
Comment (enahum) [inline plugin/hooks.go]: is your thinking here to use this endpoint to get the data of the notification in case we only send the notification id to tighten security? cause if this is the case at least for iOS a flag needs to be set in the original notification payload

File: plugin/hooks.go (hunk 8 above)
Comment (enahum) [reply plugin/hooks.go]: I'm with @hanzei here, I don't think we should care if this errors out other than logging the error. and the plugin should retry storing the ack if it fails for some reason.

File: plugin/hooks.go (hunk 16 above)
Comment (jespino) [reply plugin/hooks.go]: Yes, that was my thinking, but I'm open to other options. This is an exploratory PR that needs the mobile implementation details to makes sense entirely, so... whatever you think is better for the mobile apps, is the correct option for me :)

Comment (enahum) [reply plugin/hooks.go]: Then it does make sense, probably it will need a few more tweaks but it is the right approach

File: app/notification_push.go (hunk 14 above)
Comment (jespino) [reply app/notification_push.go]: No here, this is the `sendPushNotification` which is the async part which register the notification to be sent.

File: app/notification_push.go (hunk 15 above)
Comment (jespino) [reply app/notification_push.go]: This ID refers to the notification itself (a notification can be sent to multiple devices, and each one will receive the same `notificationId`) I'm not 100% sure if is something that we want. Actually this `notificationId` can be rename to other thing to get it less confusing, something like `notificationTrackingId`

Comment (enahum) [reply app/notification_push.go]: Every id should be unique, even if the recipient is the same user

File: api4/system.go (hunk 12 above)
Comment (jespino) [reply api4/system.go]: The `notificationId` is shared by all the devices, so... if you want to know the exact received notification you need both, the deviceId and the notificationId. We could add another generated id if it is necessary for each device notification.

File: api4/system.go (hunk 13 above)
Comment (jespino) [reply api4/system.go]: It is not needed, but the idea is to be able to modify the notification during the ACK. But maybe is not the best way to do it.

File: app/notification_push.go (hunk 15 above)
Comment (jespino) [reply app/notification_push.go]: Yes, the thing is, if we want to track before the enqueue, we need an id there which is share to each notification sent to the same user in different devices, so we can add another extra ID to each notification per device.

File: api4/system.go (hunk 12 above)
Comment (jespino) [reply api4/system.go]: Using a unique ack Id for each push notification (not sharing it between devices).

File: plugin/hooks.go (hunk 6 above)
Comment (hanzei) [reply plugin/hooks.go]: I see, thanks for the clarification.

File: plugin/hooks.go [hunk 17]
```diff
@@ -149,4 +154,41 @@ type Hooks interface {
...
+	// as response to the ACK, and the device will update the notification
+	// stored in the device.
+	//
+	// Minimum server version: 5.10
+	PushNotificationAck(c *Context, ackId string, recievedAt, ackAt int64) *model.PushNotification
```
Comment (hanzei) [inline plugin/hooks.go]: I would opt to mark this also as internal. 1/5

File: api4/system.go (hunk 13 above)
Comment (hanzei) [reply api4/system.go]: Just dropping the `return` seams fine, doesn't it?

File: app/notification_push.go [hunk 18]
```diff
@@ -270,20 +314,45 @@ func (a *App) StopPushNotificationsHubWorkers() {
...
+		pluginContext := a.PluginContext()
+		pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			updatedMsg = hooks.PushNotificationWillBeSent(pluginContext, &msg)
+			return true
+		}, plugin.PushNotificationWillBeSentId)
+	}
+	mlog.Debug("Push notification will be sent", mlog.String("notificationId", updatedMsg.Id), mlog.String("ackId", updatedMsg.AckId), mlog.String("deviceId", updatedMsg.DeviceId))
+
+	if updatedMsg == nil {
+		return nil
```
Comment (hanzei) [inline app/notification_push.go]: Maybe add a debug message here? This would help plugins devs.

Comment (jespino) [reply app/notification_push.go]: Yes, makes sense. Something like "Push notification inhibited from plugin".

File: plugin/hooks.go (hunk 17 above)
Comment (jespino) [reply plugin/hooks.go]: I don't see that, this open a lot of possibilities for plugin developers, like modify the notification on ack. For example for enrich notifications.

Comment (hanzei) [reply plugin/hooks.go]: I don't have a strong opinion on this. Fine with the current way.

File: model/push_notification.go [hunk 19]
```diff
@@ -26,7 +26,15 @@ const (
 	MHPNS = "https://push.mattermost.com"
 )
 
+type PushNotificationAck struct {
+	Id         string `json:"id"`
+	ReceivedAt int64  `json:"received_at"`
```
Comment (crspeller) [inline model/push_notification.go]: Maybe some clearer names here. `ClientRecievedAt` `ServerAckAt`?

File: api4/system.go [hunk 20]
```diff
@@ -317,3 +319,37 @@ func getRedirectLocation(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			notification = hooks.PushNotificationAck(pluginContext, ack.Id, ack.ReceivedAt, ack.AckAt)
+			return true
+		}, plugin.PushNotificationAckId)
+	}
+
+	mlog.Debug("Push notification ack received", mlog.String("ackId", ack.Id), mlog.Int64("receivedAt", ack.ReceivedAt), mlog.Int64("ackAt", ack.AckAt))
+
+	if notification != nil {
+		w.Write([]byte(notification.ToJson()))
```
Comment (crspeller) [inline api4/system.go]: We want this to enable the send nothing in the initial notification so info doesn't travel though Apple/Google feature right?

File: plugin/hooks.go (hunk 1 above)
Comment (crspeller) [reply plugin/hooks.go]: Why do you need this hook?

File: api4/system.go (hunk 20 above)
Comment (jespino) [reply api4/system.go]: Yes, exactly, that is one of the features that could be implemented as a plugin using this "Ack" notification return. Other one is workaround the 4K notifications size limit, allowing us to enrich as much as we want the push notifications.

**Comment by jasonblais** [comment]
Hey @jespino, should I freeze this PR or are there outstanding questions remaining?

**Comment by jespino** [comment]
@jasonblais actually this should be closed because @enahum is solving the underneath problem that we were trying to solve here, and we prefer to not include extensibility for now until we are sure everything is working smoothly in the push notifications part. so... I going to close it for now, and we can reopen it in some months if we think makes sense.
